plugins/node/instrumentation-amqplib/src/amqplib.ts

```diff
@@ -0,0 +1,749 @@
+            options?.correlationId,
+        },
+      }
+    );
+    const modifiedOptions = options ?? {};
+    modifiedOptions.headers = modifiedOptions.headers ?? {};
+
+    propagation.inject(
+      trace.setSpan(context.active(), span),
+      modifiedOptions.headers
+    );
+
+    return { span, modifiedOptions };
+  }
+
+  private endConsumerSpan(
+    message: InstrumentationMessage,
+    isRejected: boolean | null,
+    operation: EndOperation,
+    requeue: boolean | undefined
+  ) {
+    const storedSpan: Span | undefined = message[MESSAGE_STORED_SPAN];
+    if (!storedSpan) return;
+    if (isRejected !== false) {
+      storedSpan.setStatus({
+        code: SpanStatusCode.ERROR,
+        message:
+          operation !== EndOperation.ChannelClosed &&
+          operation !== EndOperation.ChannelError
+            ? `${operation} called on message${
+                requeue === true
+                  ? ' with requeue'
+                  : requeue === false
+                  ? ' without requeue'
+                  : ''
+              }`
+            : operation,
+      });
+    }
+    this.callConsumeEndHook(storedSpan, message, isRejected, operation);
+    storedSpan.end();
+    delete message[MESSAGE_STORED_SPAN];
+  }
+
+  private endAllSpansOnChannel(
+    channel: InstrumentationConsumeChannel,
+    isRejected: boolean,
+    operation: EndOperation,
+    requeue: boolean | undefined
+  ) {
+    const spansNotEnded: { msg: amqp.Message }[] =
+      channel[CHANNEL_SPANS_NOT_ENDED] ?? [];
+    spansNotEnded.forEach(msgDetails => {
+      this.endConsumerSpan(msgDetails.msg, isRejected, operation, requeue);
+    });
+    Object.defineProperty(channel, CHANNEL_SPANS_NOT_ENDED, {
+      value: [],
+      enumerable: false,
+      configurable: true,
+    });
+  }
+
+  private callConsumeEndHook(
+    span: Span,
+    msg: amqp.ConsumeMessage,
+    rejected: boolean | null,
+    endOperation: EndOperation
+  ) {
+    if (!this._config.consumeEndHook) return;
+
+    safeExecuteInTheMiddle(
+      () => this._config.consumeEndHook!(span, { msg, rejected, endOperation }),
+      e => {
+        if (e) {
+          diag.error('amqplib instrumentation: consumerEndHook error', e);
+        }
+      },
+      true
+    );
+  }
+
+  private checkConsumeTimeoutOnChannel(channel: InstrumentationConsumeChannel) {
+    const currentTime = new Date().getTime();
+    const spansNotEnded: { msg: amqp.Message; timeOfConsume: Date }[] =
+      channel[CHANNEL_SPANS_NOT_ENDED] ?? [];
+    let i: number;
+    for (i = 0; i < spansNotEnded.length; i++) {
+      const currMessage = spansNotEnded[i];
+      const timeFromConsume = currentTime - currMessage.timeOfConsume.getTime();
+      if (timeFromConsume < this._config.consumeTimeoutMs!) {
+        break;
+      }
+      this.endConsumerSpan(
+        currMessage.msg,
+        null,
+        EndOperation.InstrumentationTimeout,
+        true
+      );
+    }
+    spansNotEnded.splice(0, i);
```

Comment:
Is it intended to mutate `spansNotEnded` during iteration? Won't this skip some elements as `i` is not adapted accordingly.

Reply:
This is done at the end of the iteration and uses the last value of `i` as used when the for loop is `break` or end.
Do you think it's incorrect?

Reply:
seems I misread this during review, it looked like this line is inside the loop not after the loop.